### DISCUSSION TITLE
PAR request with `authorization_details` to be part of JWT.

### DISCUSSION THREAD

**Discussion by ramonpenteado** [keycloak/keycloak#22991] [Q&A] [answered]
I have read lot's of discussions around here, but none seems clear to me. So first of all, Im sorry to bring this up again in a different topic.

What we are trying to achieve is to have `authorization_details` request body param to be sent back as a claim in JWT for PAR request.

I've been reading this [discussion](https://github.com/keycloak/keycloak-community/pull/325) that seems to be adding some experimental feature, such as `dynamic_scopes` (which is a nice one), and there seems to be a topic about `authorization_details`, but I could not find any way to set it up, or if it was even implemented.
I saw this [thread](https://github.com/keycloak/keycloak/discussions/17279) as well recently updated, saying that `authorization_details` isn't something possible with Keycloack.
I understand that a custom mapper won't work, as I can't handle request body params. Not sure if I can customize anything else in this level.

So I have basically two questions:
1 - Am I missing something to get `authorization_details` to work? How can I achieve that?
2 - If this is not yet supported, are we looking to have that in a near future? 

Appreciate all that will take a moment to try and help me. Thanks a lot.

**Comment by justin-tay** [comment, marked as answer]
Can't really comment on when the feature will be there, but I don't think that it is currently there.

From what I can tell, the addition of the `dynamic_scopes` feature, when enabled, adds some processing of static and dynamic scopes that are converted to a `AuthorizationDetails` so that if the `authorization_details` was processed both the `scope` and `authorization_details` would end up in the same place internally to be processed. For instance you can't call `clientSessionCtx.getAuthorizationRequestContext()` if the feature isn't enabled, but it doesn't seem like it would contain any content from the `authorization_details` parameter even if you enabled the feature.

I don't see any place that actually processes the `authorization_details` parameter though, which means that it would be treated as an additional parameter, which depending on endpoint, may or may not be stored in the client session notes with a specific key.

If you are using PAR then your mapper should be able to get the `authorization_details` parameter that you set in the PAR request in the `client_request_param_authorization_details` client session note.

```java
@Override
protected void setClaim(final IDToken token,
                        final ProtocolMapperModel mappingModel,
                        final UserSessionModel userSession,
                        final KeycloakSession keycloakSession,
                        final ClientSessionContext clientSessionCtx) {
    clientSessionCtx.getClientSession().getNote("client_request_param_authorization_details");
} 
```

**Reply by justin-tay** [reply]
One additional thing to note is that the code in `ParEndpoint` is going to strip JSON arrays from the variable, so your parsing logic needs some way to deal with that.

https://github.com/keycloak/keycloak/blob/6830735ed9692a275208183e14c7af32c661826c/services/src/main/java/org/keycloak/protocol/oidc/par/endpoints/ParEndpoint.java#L153-L157

**Reply by ramonpenteado** [reply]
Then a custom mapper should do the trick? Is that right? Im very happy to know that actually. Basically I should enable the feature and `getNote("client_request_param_authorization_details")`.
I'll give it try. Sounds promising.
Thank you @justin-tay

**Reply by justin-tay** [reply]
You don't have to enable the dynamic scopes feature unless you want to use it.

**Reply by ramonpenteado** [reply]
It worked perfectly. Thanks a lot @justin-tay
